graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+        # outside logic controls when that is the case
+        # for example, it assumes parent is already in cluster keys of COLLAPSE node
+
+    ---------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param child: `node` with `attribute` in `column`
+    :returns: graphistry instance, new_parent_name of super cluster
+    """
+
+    ndf, edf, src, dst, node = unpack(g)
+
+    new_parent_name = get_new_node_name(ndf, parent, child)
+
+    ndf.loc[ndf[node] == parent, COLLAPSE_NODE] = new_parent_name
+    ndf.loc[ndf[node] == child, COLLAPSE_NODE] = new_parent_name
+
+    g._nodes = ndf
+    return g, new_parent_name
+
+
+def collapse_edges(g, parent, attribute, column, new_parent_name):
+    """
+        Sets new_parent_name to edge list in edf dataframe
+        Finds incluster nodes and adds those as well, but does not add it to seen_dict as that breaks traversal
+
+    --------------------------------------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param attribute: `attribute` to collapse in `column` over
+    :param column: `column` to collapse over
+    :param new_parent_name: name of collapsed node
+    """
+    # get out of cluster and in cluster nodes from parent node
+    # then feed incluster nodes to new_parent_name
+    outcluster, incluster, tdf = get_edges_in_out_cluster(g, parent, attribute, column)
+    # keep incluster nodes and assign them to new parent
+    ndf, edf, src, dst, node = unpack(g)
+
+    for node in incluster:
+        edf.loc[edf[src] == parent, COLLAPSE_SRC] = new_parent_name
+        edf.loc[edf[dst] == parent, COLLAPSE_DST] = new_parent_name
+
+        edf.loc[edf[dst] == node, COLLAPSE_DST] = new_parent_name
+        edf.loc[edf[src] == node, COLLAPSE_SRC] = new_parent_name
+
+    g._edges = edf
+    return g
+
+
+def has_property(g, ref_node, attribute, column: str) -> bool:
+    """
+        Checks if ref_node is in node dataframe in column with attribute
+
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param ref_node: `node` to check if it as `attribute` in `column`
+    :returns bool"""
+    ndf, edf, src, dst, node = unpack(g)
+    ref_node = unwrap_key(ref_node)
+    return ref_node in ndf[ndf[column] == attribute][node].values
+
+
+def check_default_columns_present_and_coerce_to_string(g):
+    """
+        Helper to set COLLAPSE columns to nodes and edges dataframe, while converting src, dst, node to dtype(str)
+
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :returns graphistry instance
+    """
+    ndf, edf, src, dst, node = unpack(g)
```

Comment:
this is a mutation on input ndf, probably should be:

```python
ndf2 = ndf.copy(deep=False) # or True if later you mutate the cells.. which you seem to...
...
g2 = g.nodes(ndf2)
...
return g2 
```

Reply:
I handle this in outer scope, in compute.py

`        from .collapse import collapse_by

        #__doc__ += collapse_by.__doc__
        g = self.bind()
        # TODO FIXME CHECK SELF LOOPS?
        return collapse_by(
            g,
            start_node=node,
            parent=node,
            attribute=attribute,
            column=column,
            seen={},
            self_edges=self_edges,
            unwrap=unwrap,
            remove_collapse=remove_collapse,
        )
`

Reply:
@silkspace  `g = self.bind()` is a shallow copy: it'll do `res._nodes = self._nodes` instead of `res._nodes = self._nodes.copy()` . Likewise, I'm fuzzy on pandas semantics around mutation, so you may even need `nodes.copy(deep=True)`.

Reply:
a test can be:

```python
g = graphistry.edges(...)
g2 = g.bind()
g2._edges[...] = something # mutate...
print(g._edges)
```